ishop/views.py

```diff
@@ -1,3 +1,179 @@
+from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.contrib.auth.views import LoginView, LogoutView
+from django.shortcuts import redirect
+from django.utils import timezone
+from django.views.generic import ListView, CreateView, UpdateView
+from django.contrib import messages
 
-# Create your views here.
+from ishop.forms import CustomUserCreationForm
+from ishop.models import Good, ShopUser, Purchase, Refund
+
+
+class SuperUserRequiredMixin(LoginRequiredMixin, UserPassesTestMixin):
+    def test_func(self):
+        return self.request.user.is_superuser
+
+
+class GoodsListView(ListView):
+    paginate_by = 10
+    http_method_names = ['post', 'get']
+    template_name = 'goods_list.html'
+    queryset = Good.objects.filter(in_stock__gt=0)
+    extra_context = {'title': 'Online shop'}
+
+    def post(self, request, *args, **kwargs):
+        if not request.user.is_authenticated:
+            msg = "Only logged users can buy"
+            messages.warning(self.request, msg)
+            return redirect('login')
+
+        user = ShopUser.objects.get(id=request.user.id)
+        good = Good.objects.get(id=self.request.POST['pk'])
+        quantity = int(self.request.POST['quantity'])
+        price = int(self.request.POST['price'])
+        in_stock = good.in_stock
+        wallet = user.wallet
+        amount = quantity * price
+
+        if wallet < amount:
+            msg = "You don't have enough money for this purchase"
+            messages.error(self.request, msg)
+            return redirect('goods')
+
+        if in_stock < quantity:
+            msg = "We don't have enough goods in stock"
+            messages.warning(self.request, msg)
+            return redirect('goods')
+
+        user.wallet -= amount
+        user.save()
+        good.in_stock -= quantity
+        good.save()
+        Purchase.objects.create(customer=user, good=good, quantity=quantity, price=price)
+
+        msg = "Your purchase is done"
+        messages.success(self.request, msg)
+
+        return self.get(request, *args, **kwargs)
+
+
+class Account(LoginRequiredMixin, ListView):
+    model = Purchase
+    paginate_by = 10
+    template_name = 'account.html'
+    extra_context = {'title': 'My account'}
+    time_interval_to_refund = 3  # in minutes
+
+    def get_queryset(self):
+        queryset = self.model.objects.filter(customer=self.request.user)
+        return queryset
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        delta = timezone.now() - timezone.timedelta(minutes=self.time_interval_to_refund)
+        context['refund_possible'] = self.model.objects.filter(customer=self.request.user, datetime__gt=delta)
+        in_refund = Refund.objects.filter(purchase__in=self.get_queryset())
+        context['in_refund'] = in_refund.values_list('purchase__pk', flat=True)
+        context['balance'] = ShopUser.objects.get(id=self.request.user.id).wallet
```

Comment:
What for???? Why not self.request.user.wallet, and that's all?

Reply:
It is because I'm not confident yet about how it works)
The reason was not to mix up logged User with AnonymusUser - but, I see, it looks like nonsense

changed now